types/next/index.d.ts

```diff
@@ -10,7 +12,43 @@
 import * as http from "http";
 import * as url from "url";
 
+import { Response as NodeResponse } from "node-fetch";
+
 declare namespace next {
+    /**
+     * Context object used in methods like `getInitialProps()`
+     * <<https://github.com/zeit/next.js/issues/1651>>
+     */
+    interface NextContext {
+        /** path section of URL */
+        pathname: string;
+        /** query string section of URL parsed as an object */
+        query: {
+            [key: string]: any
+        };
+        /** String of the actual path (including the query) shows in the browser */
+        asPath: string;
+        /** HTTP request object (server only) */
+        req?: http.IncomingMessage;
+        /** HTTP response object (server only) */
+        res?: http.ServerResponse;
+        /** Fetch Response object (client only) - from https://developer.mozilla.org/en-US/docs/Web/API/Response */
+        jsonPageRes?: NodeResponse;
+        /** Error object if any error is encountered during the rendering */
+        err?: Error;
+        /** a callback that executes the actual React rendering logic (synchronously) */
+        renderPage(
```

Comment:
As I just learned from @timneutkens, `renderPage only exists on the `ctx` for `next/document`. The rest seems the same, so maybe `interface NextDocumentContext extends NextContext`?

I'm also not sure about `jsonPageRes`. I was only inspecting the context on the server.

Reply:
@programbo Yeah, it seems like a good idea to include a base interface which can be extended for each `getInitialProps()` context types. I've done it this way initially because the basic component, `App`, and `Document` implement the context object inconsistently it's a nightmare to type on my own :')

Reply:
Cool so @programbo should I wait for some updates? :)